Add F[Boolean] syntax

BalmungSan opened the pull request:
I know that, in general, we prefer to avoid `F[Boolean]` but sometimes we have those values and I personally think this syntax can be very useful.

Let me know what you think about this change and the **Scaladoc** _(I couldn't come up with something better)_.
Also, for testing this change I could only think on a simple truth table but I am not sure if that is a good test.

rossabaker commented:
There's no clear line between the two, but a lot of syntactic convenience goes in mouse instead of cats.  Might this fit better there?

johnynek commented on core/src/main/scala/cats/syntax/monad.scala:
```diff
@@ -20,3 +24,28 @@ final class MonadIdOps[A](private val a: A) extends AnyVal {
+      case true  => F.pure(false)
+      case false => other
+    }
+}
```
This only needs Apply I think, not Monad.

BalmungSan replied:
Well, not. That is exactly the point of this addition.

Right now people can just do `(fa, fb).mapN(_ && _)` but that will evaluate `fb` even when it is not needed, if you need to preserve that _"laziness"_ then you to do this exactly `flatMap` which, IMHO, hides the intention.

Daenyth replied:
That makes it no longer short-circuit which is the same kind of footgun that ifA has

johnynek replied:
My comment got moved. This comment was on xor but my review and your push crossed paths.

XOR only needs Apply. I understand flatMap on the others.

BalmungSan replied:
@johnynek ah right, my bad, sorry.
Yeah, I removed `xor` exactly for that.

BalmungSan commented:
I renamed `and` & `or` to `andM` & `orM` per recommendation from Rob in the **discord** server.

I also removed `xor` since I think the main selling point of this addition is preserving the laziness that most folks expect of `and` & `or` which you lose with the typical advice of `(fa, fb).mapN(_ && _)`

johnynek commented on core/src/main/scala/cats/syntax/monad.scala:
```diff
@@ -20,3 +24,28 @@ final class MonadIdOps[A](private val a: A) extends AnyVal {
+  /**
+   * Behaves like `||` but inside the `F` context.
+   *
+   * Wont evaluate `other` unless this evaluates to `false`
```
This is wrong. Also underscores the need for tests.

BalmungSan replied:
Care to elaborate why that is wrong?

johnynek replied:
Or returns true if either are true. This returns false if the first is true.

BalmungSan replied:
ups... that happens for copypasting code 😅
I am sorry for that.

Yeah I agree tests are good, but then again how should I test them? The truth tables?

johnynek replied:
I would say that for all pairs of inputs `Id[Boolean]` ops should behave like the normal ops on Boolean, which yes is equivalent to testing the truth table.

johnynek commented on core/src/main/scala/cats/syntax/monad.scala:
```diff
@@ -20,3 +24,28 @@ final class MonadIdOps[A](private val a: A) extends AnyVal {
 }
+
+final class MonadBooleanOps[F[_]](private val fa: F[Boolean]) extends AnyVal {
+
```
She we also add nand, nor, any, all?

BalmungSan replied:
Since it seems that `and` and `or` are already niche enough that most folks seem to be against this addition, I doubt there is value on those others.

IMHO, these two are useful enough to be on **cats** but that is up to the consensus of everyone.

danicheg reviewed (commented):
I agree with @rossabaker that https://github.com/typelevel/mouse is a probably better place for that syntax. `Mouse` already has some `F[Foo]` syntax, so it would be natural to have `F[Boolean]` syntax too.

BalmungSan commented:
Moved to **mouse** based on feedback.
https://github.com/typelevel/mouse/pull/297
